[Sketch size] Reduce sketch size if needed

clumsy-stefan opened the pull request:
This PR is not (yet) intended to be included in the mega branch. It's intended currently for (other) selfcompilers who also have difficulties to get all their wanted plugins into one sketch, especially for 1M nodes which have to use the 2-step OTA.
All changes can be included/excluded via simple #define statements in Custom.h. See the commit messages what exactly changed in the individual commits. Standard build is not affected as long as there are no additional #defines done!
If I find time I'll probably add more commits to this. Any comments, feedback, suggestions and commits are welcome!

clumsy-stefan committed:
[1M minimal Size] Undefine graphical meminfodetail

Undefine graphical meminfodetail if BUILD_MINIMAL_OTA is defined to reduce sketch size.
Reduces sketch about 3k.

clumsy-stefan committed:
[1M minimal Size] Exclude all debug logs from standard files

Exclude all >= LOG_LEVEL_DEBUG except in Controllers, Plugins and Notifications.
Reduces sketch about 3k.

TD-er commented:
You may add this `BUILD_NO_DEBUG` to the minimal build defines.
Then it will also be checked by Travis.

clumsy-stefan commented:
I didn't want to interfere with the current builds/config's, so that's why I left it out... but if it's ok for you I'll add a commit for `define_plugin_sets.h` and include the `BUILD_NO_DEBUG` there...

TD-er commented:
Yep, as long as this isn't merged, it will not interfere at all 
And on the other hand, if you're touching a lot of files, it should be merged not too long from now. So then it should be tested first by at least building it via Travis.

clumsy-stefan committed:
[1M minimal Size] Include BUILD_NO_DEBUG in plugin sets

Add BUILD_NO_DEBUG in define_plugin_sets.h to PLUGIN_BUILD_MINIMAL_OTA.
This makes commit #316691ad07731798687ada096907c5ccb61fb1b6 included and active in 1M minimal OTA build.

clumsy-stefan committed:
[1M minimal Size] Remove unused variable..

...and make travis compiler happy again!

clumsy-stefan committed:
[1M minimal Size] Remove another unused variable..

I have no clue why my compiler does not complain about these, even though I have -Werror=unused-variable in the command line options...

clumsy-stefan commented:
Done.... I hate my compiler... it does neither warn nor throw an error on unused variables... no clue why... but it seems to be OK now ;)

TD-er commented:
And that's why I wanted Travis to have to take a look too ;)

clumsy-stefan commented:
sure, but the resulting size is exactly the same, no matter if the unused variables are included or not... will probably have an impact on memory though...

TD-er commented:
If you exclude some (flash) strings, it will decrease in binary size.
But I've seen these decreases to be in larger jumps.
So removing 10 strings can seem to have no effect and then removing an extra single byte can make the difference you aimed for.

TD-er commented:
The last build made by Travis of your commit:
```
Checking size .pioenvs/minimal_ESP8285_1024_OTA/firmware.elf
Memory Usage -> http://bit.ly/pio-memory-usage
DATA:    [=====     ]  49.5% (used 40584 bytes from 81920 bytes)
PROGRAM: [=======   ]  67.3% (used 601300 bytes from 892912 bytes)
```

And the filesize of the 20190116 minimal build: 614400 bytes
So that's about 13k less.

clumsy-stefan commented:
that's not too bad I think... the plattformIO compiler seems to be quite a bit more efficient than the Arduino IDE one...

TD-er commented:
Don't forget to do a full (clean) rebuild every now and then.
It may still use pre-compiled object files, which not only do not help reducing the filesize, but also may lead to undefined behavior sometimes.

clumsy-stefan committed:
[1M minimal Size] Exclude debug from minimal Controllers/Plugins

Exclude debug logs from C001, C002, C005, C008, C009, C013 and P001, P026, P033 which are compiled  in by default in the PLUGIN_BUILD_MINIMAL_OTA set.
Reduces sketch about 1k.
